ext/cl_khr_command_buffer.asciidoc

```diff
@@ -0,0 +1,1875 @@
+
+
+==== Add new section "Section 5.X.1 - Command Buffer Lifecycle"
+
+A command-buffer is always in one of the following states:
+
+[[recording]]
+Recording:: Initial state of a command-buffer on creation, where commands can be
+recorded to the command-buffer.
+
+[[executable]]
+Executable:: State after command recording has finished with
+*clFinalizeCommandBufferKHR* and the command-buffer may be enqueued.
+
+[[pending]]
+Pending:: Once a command-buffer has been enqueued to a command-queue it enters
+the Pending state until completion, at which point it moves back to the
+<<executable, Executable>> state.
+
+[[invalid]]
+Invalid:: A command-buffer can enter the Invalid state if a resource that was
+used in a command has been modified or freed. The only valid operation to
+perform on a command-buffer in the Invalid state is to call
+*clReleaseCommandBufferKHR* for each of the reference counts the application
+owns.
+
+image::images/commandbuffer_lifecycle.svg[align="center", title="Lifecycle of a command-buffer."]
+
+[[pending_count]]
+The Pending Count is the number of copies of the command
+buffer in the <<pending, Pending>> state. By default a command-buffer's Pending
+Count must be 0 or 1. If the command-buffer was created with
+`CL_COMMAND_BUFFER_SIMULTANEOUS_USE_KHR` then the command-buffer may have a
+Pending Count greater than 1.
+
+==== Add new section "Section 5.X.2 - Creating Command Buffer Objects"
+
+The function
+indexterm:[clCreateCommandBufferKHR]
+[source]
+----
+cl_command_buffer clCreateCommandBufferKHR(
+    cl_uint num_queues,
+    const cl_command_queue* queues,
+    const cl_command_buffer_properties_khr* properties,
+    cl_int* errcode_ret);
+----
+Is used to create a command-buffer that can record commands to the specified
+queues.
+
+[NOTE]
+====
+Upon creation the command-buffer is defined as being in the
+<<recording, Recording>> state, in order for the command-buffer to be enqueued
+it must first be finalized using *clFinalizeCommandBufferKHR* after which no
+further commands can be recorded. A command-buffer is submitted for execution
+on command-queues with a call to *clEnqueueCommandBufferKHR*.
+====
+
+_num_queues_ The number of command-queues listed in _queues_. This extension
+only supports a single command-queue, so this **must** be one.
+
+_queues_  Is a pointer to a command-queue that the command-buffer commands will
+be recorded to. _queues_ must be a non-`NULL` value.
+
+_properties_  Specifies a list of properties for the command-buffer and their
+corresponding values. Each property name is immediately followed by the
+corresponding desired value. The list is terminated with 0.
+The list of supported properties is described in the table below. If a
+supported property and its value is not specified in properties, its
+default value will be used. _properties_ can be `NULL` in which case the
+default values for supported command-buffer properties will be used.
```

Comment:
I don't think this is wrong per se, though it is a little odd and I agree it is semi-recursive.

FWIW, command queue properties has a similar issue, and it's "solved" by having `cl_command_queue_properties` be the bitfield type specifying the actual property bits and `cl_queue_properties` be the pointer to the property-value pairs passed to `clCreateCommandQueueWithProperties`.  This always seemed a little icky and I can never remember which type gets used for which purpose.  I wonder if we would have been better off with just `cl_command_queue_properties` and a similar semi-recursive usage.

Reply:
When I based this off of `clCreateCommandQueueWithProperties`, I completely missed that `cl_queue_properties` != `cl_command_queue_properties` 🤦 So the semi-recursive relationship wasn't intentional. I think my preference would be for two separate types but not sure if that's too significant a change at this point.

Reply:
> I don't think this is wrong per se

> I think my preference would be for two separate types but not sure if that's too significant a change at this point.

Stepping back for a moment, `cl_bitfield` and `cl_properties` are all just a `cl_ulong`, so this may be more trivial that I was thinking.

Reply:
I think I moved the changes we discussed over correctly, please check https://github.com/KhronosGroup/OpenCL-Docs/pull/711/commits/5dda69bfb85d13f7a86d4f847b6036b70a6b7a96.

Reply:
change LGTM, updated headers PR to reflect it in https://github.com/KhronosGroup/OpenCL-Headers/pull/182/commits/eafab6a1c77007d30f334db475949cf983009307